server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1275,11 +1308,35 @@ public DeleteResult delete(Delete delete) throws IOException {
                 e.addSuppressed(inner);
             }
             throw e;
+        } finally {
+            adjustReservingDocs(-reservedDocs);
         }
         maybePruneDeletes();
         return deleteResult;
     }
 
+    private Exception tryReserveDocs(Operation operation, int addingDocs) {
+        assert operation.origin() == Operation.Origin.PRIMARY : operation;
+        assert operation.seqNo() == SequenceNumbers.UNASSIGNED_SEQ_NO : operation;
+        final long totalDocs = indexWriter.getPendingNumDocs() + reservingNumDocs.addAndGet(addingDocs);
+        if (totalDocs > maxDocs) {
+            adjustReservingDocs(-addingDocs);
+            return new IllegalArgumentException("Number of documents in the index can't exceed [" + maxDocs + "]");
+        } else {
+            return null;
+        }
+    }
+
+    private void adjustReservingDocs(int numDocs) {
+        assert numDocs <= 0 : numDocs;
```

Comment:
Since we only ever use this to release the docs, how about renaming this to something like `releaseInFlightDocs ` and accepting a positive parameter instead?

Reply:
Good suggestion. I pushed https://github.com/elastic/elasticsearch/pull/63273/commits/294fcb478194df3b0ab9f659e42a9946ad5ab519.